miasm/os_dep/linux/environment.py

```diff
@@ -183,6 +183,9 @@ def __init__(self, number, family, type_, protocol):
         self.type_ = type_
         self.protocol = protocol
 
+    def read(self, count):
+        return b""
+
```

Comment:
I like the idea, but it seems hard to subclass, it means you have to implement a subclass of `FileDescriptorSocket`, `Network` and `LinuxEnvironment`, and make all this work together, right ?

Reply:
Kind of. It will give something like:
```python

class CustomFileDescriptorSocket(FileDescriptorSocket):
    def read(self, count):
        print("Turlututu")

class CustomNetworking(Networking):
    def socket(self, family, type_, protocol):
        fd = self.linux_env.next_fd()
        fdesc = CustomFileDescriptorSocket(fd, family, type_, protocol)
        self.linux_env.file_descriptors[fd] = fdesc
        return fd


class CustomLinuxEnvironment(LinuxEnvironment):
    def __init__(self):
        super(CustomLinuxEnvironment, self).__init__()
        self.network = CustomNetworking(self)
```

But maybe there is better: we could modify those classes to have a class variable which embed their needs. For example, for Networking:

```python
class Networking(object):
    """Network abstraction"""

    fd_generator = FileDescriptorSocket
    def __init__(self, linux_env):
        self.linux_env = linux_env

    def socket(self, family, type_, protocol):
        fd = self.linux_env.next_fd()
        fdesc = self.fd_generator(fd, family, type_, protocol)
        self.linux_env.file_descriptors[fd] = fdesc
        return fd

```

So the "overhead" may just be:
```python
class CustomNetworking(Networking):
    fd_generator = CustomFileDescriptorSocket
```


But I am not really sure if this is a suitable python pattern.
Or maybe `Networking` should take it's generator as `init` argument ?
It's a problem we already face in the `SandBox` object, which depends on os, arch, ...

@commial @p-l- I am interested if you have some feed on this.

Reply:
I'm not sure this is the same problem than the multiple inheritance of Sandbox. IMHO, it looks like more what we've done in `Jitcore` with `Cgen` or `SymbExecClass`, which reflects your last proposal.

In my opinion, the question is "what we want to provides, and what customization should be reasonably easy to implements?".
I agree with the fact that it should be easy to modify what the socket returns, its state, etc. i'm not sure that the more global Networking part needs that kind of customization possibility.

A pattern we can use would be to provides a kind of "socket factory" (sorry for this word, but it is what it is) that the Network would use to creates its sockets.
It could be a function, taking as input the socket parameters and returning an instance with the socket "interface", ie. a subclass of the socket fd.
It could also be a class, taking as `__init__` these parameters, and asked just after for successful creation or not (to keep the possibility to easiliy deny socket creation). I rather prefer the function solution, as it could be easier to return default implementation or several socket families implementation.

This "factory function" is then an attribute of the Networking class, and could be replaced with a dedicated function / property.

If this pattern become more frequent for the Linux kernel stub implementation, we could have a "config-like" class containing several factories functions, or hooks.

What do you think?

Reply:
So the socket factory would be an attribute of Network ?

Using it would be something like :

```py
class CustomFileDescriptorSocket(FileDescriptorSocket):
    def read(self, count):
        print("Chapeau pointu")

env = environment.LinuxEnvironment_x86_32()
env.network.socket_class = CustomFileDescriptorSocket
```

Is that correct ?